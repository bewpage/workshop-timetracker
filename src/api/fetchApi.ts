import { ClientRequestArgs } from 'http';
import { OperationTaskType, TaskType } from '../components/App/App';

export const options: ClientRequestArgs = {
  method: 'GET',
  headers: {
    Authorization: process.env.REACT_APP_API_KEY,
    'Content-Type': 'application/json',
  },
};

const baseAPI = async (url: string, options: any): Promise<any> => {
  const response = await fetch(process.env.REACT_APP_API_HOST + url, options);
  return await response.json();
};

export const getTasksAPI = async (
  options: any,
  url: string = '/api/tasks'
): Promise<any> => {
  // const url = '/api/tasks';
  return await baseAPI(url, options);
};

const getOperationAPI = async (
  options: any,
  id: string,
  url: string = `/api/tasks/${id}/operations`
): Promise<any> => {
  // const url = `/api/tasks/${id}/operations`;
  return await baseAPI(url, options);
};

const createTaskAPI = async (payload: TaskType): Promise<any> => {
  const option = {
    ...options,
    method: 'POST',
    body: JSON.stringify(payload),
  };
  return await getTasksAPI(option);
};

const updateTaskAPI = async (id: string, payload: TaskType): Promise<any> => {
  const option = {
    ...options,
    method: 'PUT',
    body: JSON.stringify(payload),
  };
  return await getTasksAPI(option, `/api/tasks/${id}`);
};

const deleteTaskAPI = async (id: string): Promise<any> => {
  const option = {
    ...options,
    method: 'DELETE',
  };
  return await getTasksAPI(option, `/api/tasks/${id}`);
};

const createOperationAPI = async (id: string, payload: any): Promise<any> => {
  const option = {
    ...options,
    method: 'POST',
    body: JSON.stringify(payload),
  };
  return await getOperationAPI(option, id);
};

const deleteOperationAPI = async (id: string): Promise<any> => {
  const option = {
    ...options,
    method: 'DELETE',
  };
  return await getOperationAPI(option, id, `/api/operations/${id}`);
};

const updateOperationAPI = async (
  id: string,
  payload: OperationTaskType
): Promise<any> => {
  const option = {
    ...options,
    method: 'PUT',
    body: JSON.stringify(payload),
  };
  return await getOperationAPI(option, id, `/api/operations/${id}`);
};

export const api = {
  getTasks: () => getTasksAPI(options),
  createTask: (payload: TaskType) => createTaskAPI(payload),
  updateTask: (id: string, payload: TaskType) => updateTaskAPI(id, payload),
  deleteTask: (id: string) => deleteTaskAPI(id),
  getOperations: (id: string) => getOperationAPI(options, id),
  createOperation: (id: string, payload: OperationTaskType) =>
    createOperationAPI(id, payload),
  updateOperation: (id: string, payload: OperationTaskType) =>
    updateOperationAPI(id, payload),
  deleteOperation: (id: string) => deleteOperationAPI(id),
};
