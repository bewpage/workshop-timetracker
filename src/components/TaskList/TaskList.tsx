import React, { JSX, useContext } from 'react';
import { TaskContext } from '../../states/TasksContext';
import { TaskType } from '../App/App';
import TaskListItem from '../TaskListItem/TaskListItem';
import { api } from '../../api/fetchApi';
import { ReducerEnumActions } from '../../states/taskReducer';

const TaskList = (): JSX.Element => {
  const { data, dispatch } = useContext(TaskContext);

  const deleteTaskHandler = (id: string): void => {
    // fetch api with delete method
    api.deleteTask(id).then(data => {
      if (!data.error) {
        dispatch({ type: ReducerEnumActions.deleteTask, payload: id });
      }
    });
  };

  const closeTaskHandler = (id: string): void => {
    // fetch api with update method
    const updateTask = data.tasks.find(task => task.id === id);

    if (updateTask) {
      const payload = {
        title: updateTask.title,
        description: updateTask.description,
        status: 'close',
      } as TaskType;

      // update the state API call
      api
        .updateTask(id, payload)
        .then(data => {
          if (!data.error) {
            return data;
          }
        })
        .then(response => {
          // response from server with updated task
          dispatch({
            type: ReducerEnumActions.updateTask,
            payload: response.data,
          });
        })
        .catch(error => {
          console.log(error);
        });
    }
  };

  // add new operation to task
  const addOperationFormSubmit = (
    e: React.FormEvent<HTMLFormElement>,
    id: string
  ): void => {
    e.preventDefault();

    // get data from form
    const operationForm = e.target as HTMLFormElement;
    const formData = new FormData(operationForm);
    const payload: any = {
      timeSpent: 0, // default value
    };
    formData.forEach((value, key) => {
      payload[key] = value;
    });

    // fetch api with update method
    api
      .createOperation(id, payload)
      .then(data => {
        if (!data.error) {
          return data;
        }
      })
      .then(response => {
        // response from server with new operation
        dispatch({
          type: ReducerEnumActions.addOperation,
          payload: response.data,
        });
      })
      .catch(error => {
        // TODO handle error
        console.log(error);
      });
    // clear form
    formData.forEach((value, key) => {
      operationForm[key].value = '';
    });
  };

  const taskCardRender = (value: TaskType): JSX.Element => (
    <section className="card mt-5 shadow-sm" key={value.id}>
      <div className="card-header d-flex justify-content-between align-items-center">
        <div>
          <h5>{value.title}</h5>
          <h6 className="card-subtitle text-muted">{value.description}</h6>
        </div>
        <div>
          {value.status === 'open' ? (
            <button
              className="btn btn-dark btn-sm"
              onClick={() => closeTaskHandler(value.id)}>
              Finish
            </button>
          ) : null}
          <button
            className="btn btn-outline-danger btn-sm ml-2"
            onClick={() => deleteTaskHandler(value.id)}>
            Delete
          </button>
        </div>
      </div>
      <ul className="list-group list-group-flush">
        {value.operations?.length > 0
          ? value.operations.map(operation => {
              return (
                <TaskListItem
                  key={operation.id}
                  operation={operation}
                  status={value.status}
                />
              );
            })
          : null}
      </ul>
      {/* TODO check if possible to create new component */}
      {value.status === 'open' ? (
        <div className="card-body">
          <form onSubmit={e => addOperationFormSubmit(e, value.id)}>
            <div className="input-group">
              <label htmlFor="operationDescription" className="sr-only"></label>
              <input
                id="operationDescription"
                type="text"
                placeholder="Operation description"
                className="form-control"
                name="description"
                minLength={5}
                required
              />
              <div className="input-group-append">
                <button className="btn btn-info" type="submit">
                  Add
                </button>
              </div>
            </div>
          </form>
        </div>
      ) : null}
    </section>
  );

  return (
    <>
      {data.tasks.length > 0
        ? data.tasks.map(task => {
            return taskCardRender(task);
          })
        : []}
    </>
  );
};

export default TaskList;
