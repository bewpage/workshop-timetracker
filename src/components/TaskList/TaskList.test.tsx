import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TaskList from './TaskList';
import TaskContext from '../../states/TasksContext';
import {
  mockAPIData,
  operationMockData,
  mockAPIDataTest,
} from '../../states/mocks/taskMockData';
import { api } from '../../api/fetchApi';

describe('TaskList', () => {
  beforeEach(() => {
    jest.spyOn(api, 'getTasks').mockResolvedValue({ data: [mockAPIDataTest] });
    jest
      .spyOn(api, 'getOperations')
      .mockResolvedValue({ data: [operationMockData] });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders TaskList component with data', async () => {
    const { container } = render(
      <TaskContext>
        <TaskList />
      </TaskContext>
    );

    await screen.findByText('Task Title Test');

    const taskListTitle = screen.getByText('Task Title Test');
    const taskListDescription = screen.getByText('Task Description Test');
    const finishButton = screen.getByRole('button', { name: 'Finish' });

    expect(taskListTitle).toBeInTheDocument();
    expect(taskListDescription).toBeInTheDocument();
    expect(finishButton).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders TaskList component for close task', async () => {
    jest
      .spyOn(api, 'getTasks')
      .mockResolvedValue({ data: [{ ...mockAPIDataTest, status: 'close' }] });
    const { container } = render(
      <TaskContext>
        <TaskList />
      </TaskContext>
    );

    await screen.findByText('Task Title Test');

    const taskListTitle = screen.getByText('Task Title Test');
    const taskListDescription = screen.getByText('Task Description Test');
    const finishButton = screen.queryAllByRole('button', { name: 'Finish' });
    const deleteButton = screen.getByRole('button', { name: 'Delete' });

    const addButton = screen.queryAllByRole('button', { name: 'Add' });
    const inputAdd = screen.queryAllByPlaceholderText('Operation description');

    expect(taskListTitle).toBeInTheDocument();
    expect(taskListDescription).toBeInTheDocument();
    expect(finishButton.length).toBe(0);
    expect(deleteButton).toBeInTheDocument();
    // assertion that task adds operation form is not visible for close task
    expect(addButton).toHaveLength(0);
    expect(inputAdd).toHaveLength(0);

    expect(container).toMatchSnapshot();
  });

  test('click finish task and close task', async () => {
    const closeTaskAPISpy = jest
      .spyOn(api, 'updateTask')
      .mockResolvedValue({ data: { ...mockAPIData, status: 'close' } });

    const { container } = render(
      <TaskContext>
        <TaskList />
      </TaskContext>
    );

    await screen.findByText('Task Title Test');

    expect(
      screen.getByText('Description Operation Task Test')
    ).toBeInTheDocument();

    const finishButton = screen.getByRole('button', { name: 'Finish' });
    const deleteButton = screen.getAllByRole('button', { name: 'Delete' });
    const add15mButton = screen.getByRole('button', { name: '+15m' });

    expect(deleteButton).toHaveLength(2);
    expect(add15mButton).toBeInTheDocument();
    expect(finishButton).toBeInTheDocument();

    // click on finish button
    await act(async () => {
      userEvent.click(finishButton);
    });

    // check if page is updated
    expect(closeTaskAPISpy).toHaveBeenCalledTimes(1);

    await screen.findByText('Task Title Test');

    expect(screen.queryAllByRole('button', { name: '+15m' })).toHaveLength(0);
    expect(screen.queryAllByRole('button', { name: 'Delete' })).toHaveLength(1);
    expect(container).toMatchSnapshot();
  });
});
