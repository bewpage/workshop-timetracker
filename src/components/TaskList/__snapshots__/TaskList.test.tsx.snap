// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList renders TaskList component for close task 1`] = `
<div>
  <section
    class="card mt-5 shadow-sm"
  >
    <div
      class="card-header d-flex justify-content-between align-items-center"
    >
      <div>
        <h5>
          Task Title Test
        </h5>
        <h6
          class="card-subtitle text-muted"
        >
          Task Description Test
        </h6>
      </div>
      <div>
        <button
          class="btn btn-outline-danger btn-sm ml-2"
        >
          Delete
        </button>
      </div>
    </div>
    <ul
      class="list-group list-group-flush"
    />
  </section>
</div>
`;

exports[`TaskList renders TaskList component with data 1`] = `
<div>
  <section
    class="card mt-5 shadow-sm"
  >
    <div
      class="card-header d-flex justify-content-between align-items-center"
    >
      <div>
        <h5>
          Task Title Test
        </h5>
        <h6
          class="card-subtitle text-muted"
        >
          Task Description Test
        </h6>
      </div>
      <div>
        <button
          class="btn btn-dark btn-sm"
        >
          Finish
        </button>
        <button
          class="btn btn-outline-danger btn-sm ml-2"
        >
          Delete
        </button>
      </div>
    </div>
    <ul
      class="list-group list-group-flush"
    />
    <div
      class="card-body"
    >
      <form>
        <div
          class="input-group"
        >
          <label
            class="sr-only"
            for="operationDescription"
          />
          <input
            class="form-control"
            id="operationDescription"
            minlength="5"
            placeholder="Operation description"
            type="text"
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-info"
            >
              Add
            </button>
          </div>
        </div>
      </form>
    </div>
  </section>
</div>
`;
